### ISSUE TITLE
A couple of validation errors only detected after optimization

### ISSUE THREAD

**Issue by lqd** [WebAssembly/binaryen#738]
Here are the couple of tests we found in mir2wasm. It looks like validation errors are discovered after optimization only.

I've added the C traces, but since I wasn't in a position to compile them, I also added both the non-optimized ("debug") and optimized ("release") generated wast files as well, in case the traces don't compile or something.

1) apparently a return i32 in a block while the function returns void
C trace: https://gist.github.com/lqd/b1572fdfa97fd3cc2c9bcbffdf76512e
debug wast: https://gist.github.com/lqd/e3a72ec93e7e51bfa171677b794a7784
release wast: https://gist.github.com/lqd/d2db11c78d1264f6ea4e79011e822acd

2) the validation error seems related to a function's type returning i32, while the code doesn't and is being optimized out to a (nop)
C trace: https://gist.github.com/lqd/3d0b757faffca078255f5bb8dc797d86
debug wast: https://gist.github.com/lqd/742bef11c8187c2e8a3e84f21baa5bff
release wast: https://gist.github.com/lqd/5860c431099db9bb89a65d203cbf0202

**Comment by kripken** [comment]
From the trace it looks like you aren't validating the module before optimizing it? That might explain this. We don't validate automatically, normally, as it's a costly operation. It's best to validate when the module is complete, before doing anything with it like optimizing or saving.

**Comment by lqd** [comment]
To me the incomprehension wasn't about the optimizations, but that validation passes on those seemingly incorrect wast files without optimizations ?

To showcase this further, I've updated the traces to do 2 validations, one before and one after optimizing. The validation on the unoptimized module will not detect errors, the validation on the now optimized module will print the errors out.

**Comment by kripken** [comment]
1. Looking at the new trace, something seems wrong - there is no call to BinaryenModuleCreate. Is it possible you're not calling it? I doublechecked and our tracing code should emit a call to it if you did (on current binaryen master).
2. Those wasts do fail validation on binaryen master, e.g. the first hits

````
[wasm-validator error in function $__wasm_start] unexpected false: can only drop a valid value, on 
(drop
  (call $main)
)
````

Perhaps you're not on latest master, and we added that check later?
